src/generic/stage2_build_tape.h

```diff
@@ -10,87 +10,83 @@
 #define GOTO_CONTINUE()                                                        \
   {                                                                            \
-    if (pj.ret_address[depth] == 'a') {                                        \
+    if (pj.ret_address[depth]) {                                               \
       goto array_continue;                                                     \
-    } else if (pj.ret_address[depth] == 'o') {                                 \
+    } else  {                                                                  \
       goto object_continue;                                                    \
-    } else {                                                                   \
-      goto start_continue;                                                     \
     }                                                                          \
   }
 #endif
 
+
+enum {
+  check_index_end_true = true,
+  check_index_end_false = false
+};
+
 /************
  * The JSON is parsed to a tape, see the accompanying tape.md file
  * for documentation.
  ***********/
+
+typedef struct {
+  uint32_t current_depth;
+  uint32_t current_index;
+} stage2_status;
+
+ /**
+ * Unified_machine_init is meant to be always called at least one.
+ * It sets up the state of the stage 2 parser. If there is a single
+ * atom in the JSON document, the parsing will complete.
+ * If an object or an array if found, then it will return  simdjson::SUCCESS_AND_HAS_MORE.
+ */
 WARN_UNUSED  int
-unified_machine(const uint8_t *buf, size_t len, ParsedJson &pj) {
-  uint32_t i = 0; /* index of the structural character (0,1,2,3...) */
-  uint32_t idx; /* location of the structural character in the input (buf)   */
-  uint8_t c;    /* used to track the (structural) character we are looking at,
-                   updated */
-  /* by UPDATE_CHAR macro */
-  uint32_t depth = 0; /* could have an arbitrary starting depth */
+unified_machine_init(const uint8_t *buf, size_t len, ParsedJson &pj, stage2_status& s) {
+  uint32_t depth = 0;
   pj.init();          /* sets is_valid to false          */
   if (pj.byte_capacity < len) {
     pj.error_code = simdjson::CAPACITY;
     return pj.error_code;
   }
-
+  uint32_t i = 0;
+  uint32_t idx; 
+  uint8_t c;
   /*//////////////////////////// START STATE /////////////////////////////
    */
-  SET_GOTO_START_CONTINUE()
   pj.containing_scope_offset[depth] = pj.get_current_loc();
   pj.write_tape(0, 'r'); /* r for root, 0 is going to get overwritten */
   /* the root is used, if nothing else, to capture the size of the tape */
   depth++; /* everything starts at depth = 1, depth = 0 is just for the
               root, the root may contain an object, an array or something
               else. */
+  s.current_depth = depth;
+  s.current_index = i;
   if (depth >= pj.depth_capacity) {
     goto fail;
   }
-
-  UPDATE_CHAR();
+  // we are now going to check whether we have a single atom
+  UPDATE_CHAR()
```

Comment:
A `;` after this is probably best--I understand it's a `{}` underneath, but just while reading  it looks like it's supposed to be a statement.

Reply:
Sure. Let me fix it.